src/audio/src/src.c

```diff
@@ -870,6 +823,126 @@ static int src_params(struct comp_dev *dev,
 	return err;
 }
 
+static int src_prepare_general(struct comp_dev *dev, struct comp_data *cd,
+			       struct comp_buffer __sparse_cache *source,
+			       struct comp_buffer __sparse_cache *sink)
+{
+	enum sof_ipc_frame source_format;
+	enum sof_ipc_frame sink_format;
+	int ret;
+
+	/* get source/sink data format */
+	source_format = source->stream.frame_fmt;
+	sink_format = sink->stream.frame_fmt;
+
+	ret = src_check_buffer_sizes(dev, cd, &source->stream, &sink->stream);
+	if (ret < 0)
+		goto out;
+
+	/* SRC supports S16_LE, S24_4LE and S32_LE formats */
+	if (source_format != sink_format) {
+		comp_err(dev, "src_prepare(): Source fmt %d and sink fmt %d are different.",
+			 source_format, sink_format);
+		ret = -EINVAL;
+		goto out;
+	}
+
+	switch (source_format) {
+#if CONFIG_FORMAT_S16LE
+	case SOF_IPC_FRAME_S16_LE:
+		cd->data_shift = 0;
+		cd->polyphase_func = src_polyphase_stage_cir_s16;
+		break;
+#endif /* CONFIG_FORMAT_S16LE */
+#if CONFIG_FORMAT_S24LE
+	case SOF_IPC_FRAME_S24_4LE:
+		cd->data_shift = 8;
+		cd->polyphase_func = src_polyphase_stage_cir;
+		break;
+#endif /* CONFIG_FORMAT_S24LE */
+#if CONFIG_FORMAT_S32LE
+	case SOF_IPC_FRAME_S32_LE:
+		cd->data_shift = 0;
+		cd->polyphase_func = src_polyphase_stage_cir;
+		break;
+#endif /* CONFIG_FORMAT_S32LE */
+	default:
+		comp_err(dev, "src_prepare(): invalid format %d", source_format);
+		ret = -EINVAL;
+		goto out;
+	}
+
+out:
+	if (ret < 0)
+		comp_set_state(dev, COMP_TRIGGER_RESET);
+
+	buffer_release(sink);
+	buffer_release(source);
```

Comment:
nnno, let's not do this. Formally this is correct and reduces code duplication, but I don't think it's a good coding practice. Usually you want to acquire and release or lock and unlock objects at the same level, within the same function. It's too error prone the way you do it here. Let's keep these two `buffer_release()` lines in the caller functions, or maybe you could move `buffer_acquire()` inside this function too. But let's not split it between the caller and here.